File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveFactSetRenderer.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
using System.Windows;
using System.Windows.Controls;

namespace AdaptiveCards.Rendering.Wpf
{
    public static class AdaptiveFactSetRenderer
    {
        public static FrameworkElement Render(AdaptiveFactSet factSet, AdaptiveRenderContext context)
        {
            if (factSet.Facts.Count == 0)
            {
                return null;
            }

            var uiFactSet = new Grid();
            // grid.Margin = factSet.Theme.FactSetMargins;
            uiFactSet.Style = context.GetStyle("Adaptive.FactSet");

            uiFactSet.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Auto });
            uiFactSet.ColumnDefinitions.Add(new ColumnDefinition() { Width = new GridLength(1, GridUnitType.Star) });
            int iRow = 0;

            foreach (var fact in factSet.Facts)
            {
                var uiTitle = context.Render(new AdaptiveTextBlock()
                {
                    Size = context.Config.FactSet.Title.Size,
                    Color = context.Config.FactSet.Title.Color,
                    IsSubtle = context.Config.FactSet.Title.IsSubtle,
                    Weight = context.Config.FactSet.Title.Weight,
                    Wrap = context.Config.FactSet.Title.Wrap,
                    MaxWidth = context.Config.FactSet.Title.MaxWidth,
                    Text = fact.Title
                });

                var uiValue = context.Render(new AdaptiveTextBlock()
                {
                    Size = context.Config.FactSet.Value.Size,
                    Color = context.Config.FactSet.Value.Color,
                    IsSubtle = context.Config.FactSet.Value.IsSubtle,
                    Weight = context.Config.FactSet.Value.Weight,
                    Wrap = context.Config.FactSet.Value.Wrap,
                    // MaxWidth is not applicable to the Value field of the Fact
                    // so ignore it.
                    Text = fact.Value
                });

                if (uiTitle != null || uiValue != null)
                {
                    uiFactSet.RowDefinitions.Add(new RowDefinition() { Height = GridLength.Auto });
                }

                if (uiTitle != null)
                {
                    uiTitle.Style = context.GetStyle("Adaptive.Fact.Title");
                    uiTitle.Margin = new Thickness(left: 0, top: 0, right: context.Config.FactSet.Spacing, bottom: 0);

                    Grid.SetColumn(uiTitle, 0);
                    Grid.SetRow(uiTitle, iRow);
                    uiFactSet.Children.Add(uiTitle);
                }

                if (uiValue != null)
                {
                    uiValue.Style = context.GetStyle("Adaptive.Fact.Value");

                    Grid.SetColumn(uiValue, 1);
                    Grid.SetRow(uiValue, iRow);
                    uiFactSet.Children.Add(uiValue);
                }

                if (uiTitle != null || uiValue != null)
                {
                    iRow++;
                }
            }

            return (iRow == 0) ? null : uiFactSet;
        }
    }
}

File: source/winappsdk/Renderer/lib/AdaptiveFactSetRenderer.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
#include "pch.h"

#include "AdaptiveFactSetRenderer.h"
#include "AdaptiveFactSetRenderer.g.cpp"
#include "TextHelpers.h"

using namespace Microsoft::WRL;
using namespace Microsoft::WRL::Wrappers;
using namespace ABI::AdaptiveCards::Rendering::WinUI3;
using namespace ABI::AdaptiveCards::ObjectModel::WinUI3;
using namespace ABI::Windows::Foundation;
using namespace ABI::Windows::Foundation::Collections;
using namespace ABI::Windows::UI::Xaml;
using namespace ABI::Windows::UI::Xaml::Controls;

namespace winrt::AdaptiveCards::Rendering::WinUI3::implementation
{

    rtxaml::UIElement AdaptiveFactSetRenderer::Render(rtom::IAdaptiveCardElement const& cardElement,
                                                      rtrender::AdaptiveRenderContext const& renderContext,
                                                      rtrender::AdaptiveRenderArgs const& renderArgs)
    {
        try
        {
            auto adaptiveFactSet = cardElement.as<rtom::AdaptiveFactSet>();

            rtxaml::Controls::Grid xamlGrid{};

            rtxaml::Controls::ColumnDefinition titleColumn{};
            rtxaml::Controls::ColumnDefinition valueColumn{};

            // TODO: does value(0 in this case) matter when type=Auto?
            rtxaml::GridLength factSetGridTitleLength{0, rtxaml::GridUnitType::Auto};
            rtxaml::GridLength factSetGridValueLength{1, rtxaml::GridUnitType::Star};

            titleColumn.Width(factSetGridTitleLength);
            valueColumn.Width(factSetGridValueLength);

            auto columnDefinitions = xamlGrid.ColumnDefinitions();
            columnDefinitions.Append(titleColumn);
            columnDefinitions.Append(valueColumn);

            rtxaml::GridLength factSetGridHeight{0, rtxaml::GridUnitType::Auto};

            auto heightType = cardElement.Height();

            if (heightType == rtom::HeightType::Stretch)
            {
                factSetGridHeight = {1, rtxaml::GridUnitType::Star};
            }

            auto facts = adaptiveFactSet.Facts();

            int currentFact = 0, validFacts = 0;

            for (auto fact : facts)
            {
                rtxaml::Controls::RowDefinition factRow{};
                factRow.Height(factSetGridHeight);

                auto rowDefinitions = xamlGrid.RowDefinitions();
                rowDefinitions.Append(factRow);

                auto hostConfig = renderContext.HostConfig();
                auto factSetConfig = hostConfig.FactSet();

                // Get Language
                winrt::hstring language = fact.Language();

                // Create the title xaml textblock and style it from Host options
                auto titleTextConfig = factSetConfig.Title();
                rtxaml::Controls::TextBlock titleTextBlock{};

                winrt::hstring factTitle = fact.Title();

                SetXamlInlinesWithFactSetTextConfig(renderContext, renderArgs, titleTextConfig, language, factTitle, titleTextBlock);

                // Create the value xaml textblock and style it from Host options
                auto valueTextConfig = factSetConfig.Value();
                rtxaml::Controls::TextBlock valueTextBlock{};

                winrt::hstring factValue = fact.Value();

                SetXamlInlinesWithFactSetTextConfig(renderContext, renderArgs, valueTextConfig, language, factValue, valueTextBlock);

                if (!factTitle.empty() || !factValue.empty())
                {
                    // Mark the column container with the current column

                    auto spacing = factSetConfig.Spacing();
                    titleTextBlock.Margin({0, 0, (double)spacing, 0});

                    ::AdaptiveCards::Rendering::WinUI3::XamlHelpers::SetStyleFromResourceDictionary(renderContext,
                                                                                                    L"Adaptive.Fact.Title",
                                                                                                    titleTextBlock);

                    ::AdaptiveCards::Rendering::WinUI3::XamlHelpers::SetStyleFromResourceDictionary(renderContext,
                                                                                                    L"Adaptive.Fact.Value",
                                                                                                    valueTextBlock);

                    rtxaml::Controls::Grid::SetColumn(titleTextBlock, 0);
                    rtxaml::Controls::Grid::SetRow(titleTextBlock, currentFact);

                    rtxaml::Controls::Grid::SetColumn(valueTextBlock, 1);
                    rtxaml::Controls::Grid::SetRow(valueTextBlock, currentFact);

                    // Finally add the column container to the grid, and increment the column count

                    ::AdaptiveCards::Rendering::WinUI3::XamlHelpers::AppendXamlElementToPanel(titleTextBlock, xamlGrid);
                    ::AdaptiveCards::Rendering::WinUI3::XamlHelpers::AppendXamlElementToPanel(valueTextBlock, xamlGrid);
                    ++currentFact;
                    ++validFacts;
                }
            }

            if (validFacts == 0)
            {
                // TODO: is this right? or we're supposed to do something else?
                return nullptr;
            }

            ::AdaptiveCards::Rendering::WinUI3::XamlHelpers::SetStyleFromResourceDictionary(renderContext, L"Adaptive.FactSet", xamlGrid);

            return xamlGrid;
        }
        catch (winrt::hresult_error const& ex)
        {
            // TODO: what do we do here?
            return nullptr;
        }
    }
}
